packages/react-intersection-observer/src/hooks.ts

```diff
@@ -0,0 +1,147 @@
+import {useState, useRef, useEffect, Ref} from 'react';
+
+import {isSupported} from './utilities';
+import {UnsupportedBehavior} from './types';
+
+interface Options {
+  root?: Element | string | null;
+  rootMargin?: string;
+  threshold?: number | number[];
+  children?: React.ReactNode;
+  wrapperComponent?: string;
+  unsupportedBehavior?: UnsupportedBehavior;
+}
+
+const emptyBoundingClientRect = {
+  bottom: 0,
+  height: 0,
+  left: 0,
+  right: 0,
+  top: 0,
+  width: 0,
+};
+
+export function useIntersection({
+  root,
+  rootMargin,
+  threshold,
+  unsupportedBehavior = UnsupportedBehavior.TreatAsIntersecting,
+}: Options = {}): [IntersectionObserverEntry, Ref<HTMLElement | null>] {
+  const node = useRef<HTMLElement | null>(null);
+  const lastNode = useRef<HTMLElement | null>(null);
+  const observer = useRef<IntersectionObserver | null>(null);
+  const lastObserver = useRef<IntersectionObserver | null>(null);
+
+  const [intersectionEntry, setIntersectingEntry] = useState<
+    IntersectionObserverEntry
+  >(() => ({
+    boundingClientRect: emptyBoundingClientRect,
```

Comment:
should each of these maybe be clones instead of a reference to the same object?

Reply:
I don't think it matters, all the fields are meant to be readonly